src/main/scala/com/mongodb/spark/rdd/partitioner/MongoSamplePartitioner.scala

```diff
@@ -71,35 +71,47 @@ class MongoSamplePartitioner extends MongoPartitioner {
    * @param readConfig the [[com.mongodb.spark.config.ReadConfig]]
    * @return the partitions
    */
+  // scalastyle:off cyclomatic.complexity
   override def partitions(connector: MongoConnector, readConfig: ReadConfig, pipeline: Array[BsonDocument]): Array[MongoPartition] = {
-
     Try(PartitionerHelper.collStats(connector, readConfig)) match {
       case Success(results) =>
+        val matchQuery = PartitionerHelper.matchQuery(pipeline)
```

Comment:
The sampling partitioner now uses a `count` query with the `matchQuery` to get the size of the results rather than using the `collstats` `count` value if there is a query. The `matchQuery` is added to the sampling aggregation pipeline.

If there is a `$match` filter the last partition is changed to be inclusive of the last value by using `setLastBoundaryToLessThanOrEqualTo`.